wsproto/connection.py

```diff
@@ -275,6 +275,11 @@ def events(self):
                 if frame.opcode is Opcode.PING:
                     assert frame.frame_finished and frame.message_finished
                     self._outgoing += self._proto.pong(frame.payload)
+                    yield PingReceived(frame.payload)
+
+                elif frame.opcode is Opcode.PONG:
+                    assert frame.frame_finished and frame.message_finished
+                    yield PongReceived(frame.payload)
```

Comment:
RFC 6455 is generally pretty clear about what invariants it wants implementors to enforce on their peers. There are lots of cases where it says "if your peer does this, you MUST immediately close the connection". This isn't one of them, and my guess is that this is intentional (even if the lack of all-caps might be accidental), because, well, AFAICT it's literally impossible to enforce even if we want to :-).